src/pep257.py

```diff
@@ -679,73 +678,496 @@ def __getattr__(self, item):
+        return self._run_conf
+
+    @check_initialized
+    def get_files_to_check(self):
+        """Return a generator of files and error codes to check on each file.
+
+        Walk dir trees under `self._arguments` and generate yield filnames
+        that `match` under each directory that `match_dir`.
+        The method locates the configuration for each file name and yields a
+        tuple of (filename, [error_codes]).
+
+        """
+        def _get_matches(config):
+            """Return the `match` and `match_dir` functions for `config`."""
+            match_func = re(config.match + '$').match
+            match_dir_func = re(config.match_dir + '$').match
+            return match_func, match_dir_func
+
+        for name in self._arguments:
+            if os.path.isdir(name):
+                for root, dirs, filenames in os.walk(name):
+                    config = self._get_config(root)
+                    match, match_dir = _get_matches(config)
+
+                    # Skip any dirs that do not match match_dir
+                    dirs[:] = [dir for dir in dirs if match_dir(dir)]
+
+                    for filename in filenames:
+                        if match(filename):
+                            full_path = os.path.join(root, filename)
+                            yield full_path, list(config.checked_codes)
+            else:
+                config = self._get_config(name)
+                match, _ = _get_matches(config)
+                if match(name):
+                    yield name, list(config.checked_codes)
+
+    # --------------------------- Private Methods -----------------------------
+
+    def _get_config(self, node):
+        """Get and cache the run configuration for `node`.
+
+        If no configuration exists (not local and not for the parend node),
+        returns and caches a default configuration.
+
+        The algorithm:
+        -------------
+        1.  If the current directory's configuration exists in
+            `self._cache` - return it.
+        2.  If a configuration file does not exist in this directory:
+        3.    If there's a parent directory:
+        4.      Cache it's configuration as this directory's and return it.
+        5.    Else:
+        6.      Cache a default configuration and return it.
+        7.  Else:
+        8.    Read the configuration file.
+        9.    If a parent directory exists AND the configuration file
+              allows inheritance:
+        10.     Read the parent configuration by calling this function with the
+                parent directory as `node`.
+        11.     Migrate the parent configuration with the current one and
+                cache it.
+        12. If the user has specified one of `MUTUALLY_EXCLUSIVE_OPTIONS` in
+            the CLI - return the CLI configuration with the configuration match
+            clauses
+        13. Set the `--add-select` and `--add-ignore` CLI configurations.
```

Comment:
Replace the numbers in the algorithm with bullets - it will be much easier to maintain if it changes in the future.